### PR TITLE
Refactor: KCL Lint(Rust ver.)

### PR DESCRIPTION
**Pull request by He1pa**
#### 1. Does this PR affect any open issues?(Y/N) and add issue references (e.g. "fix #123", "re #123".):

- [ ] N
- [x] Y #109  

#### 2. What is the scope of this PR (e.g. component or file name):

sema/lint
sema/resolver
ast/walker

#### 3. Provide a description of the PR(e.g. more details, effects, motivations or doc link):

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Other

Add Lint check in resolver. Main changes are:
1. mode.rs: Definition of `Linter`, the entry for lint check
2. combinedlintpass.rs: `CombinedLintPass` collects all the lints defined in the lints_def.rs
3. lints_def.rs: Defined the various lints and the corresponding lintpasses implementation
4. lintpass.rs: Definition of `Lintpass`
5. lint.rs: Definition of `Lint`
6. resolver/mode.rs: Call `lint_check_module` when recursively check the ast, and call `lint_check_scope` after check

#### 4. Are there any breaking changes?(Y/N) and describe the breaking changes(e.g. more details, motivations or doc link):

- [x] N
- [ ] Y 

#### 5. Are there test cases for these changes?(Y/N) select and add more details, references or doc links:

- [x] Unit test
- [ ] Integration test
- [ ] Benchmark (add benchmark stats below)
- [ ] Manual test (add detailed scripts or steps below)
- [ ] Other

#### 6. Release note

Please refer to [Release Notes Language Style Guide](https://kusionstack.io/docs/governance/release-policy/) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: kclvm/sema/src/lint/mod.rs [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+                set_pos!($walker, &v);
+                $walker.$method(&v.node);
+            }
+            None => (),
+        }
+    };
+}
+
+#[macro_export]
+macro_rules! set_pos {
```
Comment (Peefy) [inline kclvm/sema/src/lint/mod.rs:166]: move these to `macros.rs`?

File: kclvm/sema/src/lint/mod.rs [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+        }
+    }
+    pub fn walk_scope(&mut self, scope: &Scope) {
+        self.pass
+            .check_scope(&mut self.handler, &mut self.ctx, scope);
+    }
+}
+
+impl Resolver<'_> {
+    pub fn lint_check_module(&mut self, module: &ast::Module) {
```
Comment (Peefy) [inline kclvm/sema/src/lint/mod.rs:122]: more documents on these functions, or change it to `pub(crate)`

File: kclvm/sema/src/resolver/mod.rs [hunk 3]
```diff
@@ -57,6 +59,7 @@ impl<'ctx> Resolver<'ctx> {
             ctx: Context::default(),
             options,
             handler: Handler::default(),
+            linter: Linter::<CombinedLintPass>::new(Handler::default(), LintContext::dummy_ctx()),
```
Comment (Peefy) [inline kclvm/sema/src/resolver/mod.rs:62]: change LintContext into a internal field?

**Review by Peefy — APPROVED** [review]
LGTM.

**Review by zong-zhe — APPROVED** [review]
LGTM

**Pull request merged**